hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java

```diff
@@ -150,12 +148,12 @@ protected ClientInvocationFuture getFuture() {
 
     @Override
     public <U> CompletableFuture<U> thenApply(Function<? super V, ? extends U> fn) {
-        return future.thenApply(new DeserializingFunction<>(fn));
+        return future.thenApplyAsync(new DeserializingFunction<>(fn), defaultExecutor());
```

Comment:
shouldn't this be ran inside user thread in a callers-run-fashion?

Reply:
That was the case previously, the user-supplied callback would be executed either in the calling thread (if the future was already completed) or by the thread that completes the future (for example when the future was not completed at the time the `Function` is registered as a dependent). The problem with this approach is that the thread that completes the future can be a Hazelcast response thread and allowing arbitrary blocking user code to be executed on hazelcast threads can result in halting Hazelcast. For example the following function executed from a client-side input thread may fail to complete:
```
IMap<Integer, Integer> map = hz.getMap("test");
map.put(1, 2);
map.put(2, 4);
CompletionStage<Integer> stage = map.getAsync(1).thenApply(v -> {
    System.out.println(Thread.currentThread()); // never completes when executed from hz.client_0.IO.thread-in-*
    return map.get(v);
});
```

`CompletionStage` [javadoc](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/CompletionStage.html) reads:
> Additionally, execution of a new stage's computations may be arranged in any of three ways: default execution, default asynchronous execution (using methods with suffix async that employ the stage's default asynchronous execution facility), or custom (via a supplied Executor). The execution properties of default and async modes are specified by CompletionStage implementations, not this interface.

So I think that this change is aligned with `CompletionStage` contract and will save users from risk of blocking critical hazelcast threads.